osu.Game/Database/EFToRealmMigrator.cs

```diff
@@ -30,6 +34,171 @@ public void Run()
+                {
+                    foreach (var beatmapSet in existingBeatmapSets)
+                    {
+                        var realmBeatmapSet = new BeatmapSetInfo
+                        {
+                            OnlineID = beatmapSet.OnlineID ?? -1,
+                            DateAdded = beatmapSet.DateAdded,
+                            Status = beatmapSet.Status,
+                            DeletePending = beatmapSet.DeletePending,
+                            Hash = beatmapSet.Hash,
+                            Protected = beatmapSet.Protected,
+                        };
+
+                        migrateFiles(beatmapSet, realm, realmBeatmapSet);
+
+                        foreach (var beatmap in beatmapSet.Beatmaps)
+                        {
+                            var realmBeatmap = new BeatmapInfo
+                            {
+                                DifficultyName = beatmap.DifficultyName,
+                                Status = beatmap.Status,
+                                OnlineID = beatmap.OnlineID ?? -1,
+                                Length = beatmap.Length,
+                                BPM = beatmap.BPM,
+                                Hash = beatmap.Hash,
+                                StarRating = beatmap.StarRating,
+                                MD5Hash = beatmap.MD5Hash,
+                                Hidden = beatmap.Hidden,
+                                AudioLeadIn = beatmap.AudioLeadIn,
+                                StackLeniency = beatmap.StackLeniency,
+                                SpecialStyle = beatmap.SpecialStyle,
+                                LetterboxInBreaks = beatmap.LetterboxInBreaks,
+                                WidescreenStoryboard = beatmap.WidescreenStoryboard,
+                                EpilepsyWarning = beatmap.EpilepsyWarning,
+                                SamplesMatchPlaybackRate = beatmap.SamplesMatchPlaybackRate,
+                                DistanceSpacing = beatmap.DistanceSpacing,
+                                BeatDivisor = beatmap.BeatDivisor,
+                                GridSize = beatmap.GridSize,
+                                TimelineZoom = beatmap.TimelineZoom,
+                                Countdown = beatmap.Countdown,
+                                CountdownOffset = beatmap.CountdownOffset,
+                                MaxCombo = beatmap.MaxCombo,
+                                Bookmarks = beatmap.Bookmarks,
+                                Ruleset = realm.Find<RulesetInfo>(beatmap.RulesetInfo.ShortName),
+                                Difficulty = new BeatmapDifficulty(beatmap.BaseDifficulty),
+                                Metadata = new BeatmapMetadata
+                                {
+                                    Title = beatmap.Metadata.Title,
+                                    TitleUnicode = beatmap.Metadata.TitleUnicode,
```

Comment:
This crashes at startup for me because it's importing protected intro beatmaps which have a null metadata:
![image](https://user-images.githubusercontent.com/1329837/149493767-7b650005-7133-44f4-9b46-04af88859d73.png)
Or something else is going on, I don't know, because I have a _lot_ of beatmap sets where at least one contained beatmap has null metadata:
![image](https://user-images.githubusercontent.com/1329837/149494478-916c423c-5e58-47e4-a7d5-e36527bcf1b0.png)
Which is significantly more beatmap sets than those where all beatmaps have metadata:
![image](https://user-images.githubusercontent.com/1329837/149494595-e263cce8-6a72-4fde-b123-2c7127e44588.png)

[client.db.zip](https://github.com/ppy/osu/files/7869436/client.db.zip)

Reply:
we'll need to account for that. set level was left null during import when it didn't match across all beatmaps.

i have a separate branch that does similar deduping in realm, but that's for another day.